frame/sudo/src/tests.rs

```diff
@@ -0,0 +1,164 @@
+			Error::<Test>::RequireSudo,
+		);
+		assert_eq!(logger::log(), vec![42u64]); 
+
+		// Controls the dispatched weight.
+		let call = Box::new(Call::Logger(LoggerCall::log(42, 1)));
+		let sudo_unchecked_weight_call = SudoCall::sudo_unchecked_weight(call, 1_000);
+		let info = sudo_unchecked_weight_call.get_dispatch_info();
+		assert_eq!(info.weight, 1_000);
+	});
+}
+
+#[test]
+fn sudo_unchecked_weight_emits_events_correctly() {
+	new_test_ext(1).execute_with(|| {
+		// Set block number to 1 because events are not emitted on block 0.
+		System::set_block_number(1);
+
+		// Should emit event to indicate success when called with the root `key` and `call` is `Ok`.
+		let call = Box::new(Call::Logger(LoggerCall::log(42, 1)));
+		assert_ok!(Sudo::sudo_unchecked_weight(Origin::signed(1), call, 1_000));
+		let expected_event = TestEvent::sudo(RawEvent::Sudid(Ok(())));
+		assert!(System::events().iter().any(|a| {a.event == expected_event}));
+	})
+}
+
+#[test]
+fn set_key_basics() {
+	new_test_ext(1).execute_with(|| {	
+		// A root `key` can change the root `key`
+		assert_ok!(Sudo::set_key(Origin::signed(1), 2));
+		assert_eq!(Sudo::key(),  2u64);
+	});
+
+	new_test_ext(1).execute_with(|| {
+		// A non-root `key` will trigger a `RequireSudo` error and a non-root `key` cannot change the root `key`.
+		assert_noop!(Sudo::set_key(Origin::signed(2), 3), Error::<Test>::RequireSudo);
+		assert_eq!(Sudo::key(),  1u64);
+	});
+}
+
+#[test]
+fn set_key_emits_events_correctly() {
+	new_test_ext(1).execute_with(|| {	
+		// Set block number to 1 because events are not emitted on block 0.
+		System::set_block_number(1);
+
+		// A root `key` can change the root `key`.
+		assert_ok!(Sudo::set_key(Origin::signed(1), 2));
+		let expected_event = TestEvent::sudo(RawEvent::KeyChanged(1));
+		assert!(System::events().iter().any(|a| {a.event == expected_event}));
+		// Double check.
+		assert_ok!(Sudo::set_key(Origin::signed(2), 4));
+		let expected_event = TestEvent::sudo(RawEvent::KeyChanged(2));
+		assert!(System::events().iter().any(|a| {a.event == expected_event}));
+	});
+}
+
+#[test]
+fn sudo_as_basics() {
+	new_test_ext(1).execute_with(|| {	
+		// A privileged function will not work when passed to `sudo_as`.
+		let call = Box::new(Call::Logger(LoggerCall::log(42, 1)));
+		assert_ok!(Sudo::sudo_as(Origin::signed(1), 2, call));
+		assert_eq!(logger::log(), vec![]); 
+
+		// A non-privileged function should not work when called with a non-root `key`.
+		let call = Box::new(Call::Logger(LoggerCall::non_privileged_log(42, 1)));
+		assert_noop!(Sudo::sudo_as(Origin::signed(3), 2, call), Error::<Test>::RequireSudo);
+		assert_eq!(logger::log(), vec![]); 
+
+		// A non-privileged function will work when passed to `sudo_as` with the root `key`.
+		let call = Box::new(Call::Logger(LoggerCall::non_privileged_log(42, 1)));
+		assert_ok!(Sudo::sudo_as(Origin::signed(1), 2, call));
+		assert_eq!(logger::log(), vec![42u64]); 
```

Comment:
One more nit here. I would make non-privileged logger also log the user who made the call.

Then we would want to test `sudo_as` not only is successful with the non-privileged logger, but also that the correct user was the one logged as making the call.

This would ultimately test the full functionality of `sudo_as`. No only do we dispatch some `ensure_signed` call, but we do so with the specified user account.

You can imagine a change to this pallet which would keep this test passing (where we hard-code some user account as the dispatch account in `sudo_as`)

Reply:
I added a logger function that logs its origin. I also realized that the value for an unset storage item of type u64/i32 is 0, which I thought is a bit of a misdirection. So I opted to store both `i` and `AccountId` in vecs so in the case of empty set it is an empty vec, which I think is more intuitive. I also refactored the logger to only use storage items and gave them a bit more descriptive names.

Reply:
you can also use an option when you want to represent "no" value. Otherwise, if storage is not an option, and there is no value, the default value is used instead.